docs/language/informal/local-inference.md

```diff
@@ -0,0 +1,500 @@
+
+Note that the least closure of a type schema is always a subtype of any type
+which matches the schema, and the greatest closure of is always a supertype of
+any type which matches the schema.
+
+
+
+### Type variable elimination (least and greatest closure of a type)
+
+We define the least closure of a type `M` with respect to a set of type
+variables `T0, ..., Tn` to be `M` with every covariant occurrence of `Ti`
+replaced with `Null`, and every contravariant occurrence of `Ti` replaced
+with `Object`.
+
+We define the greatest closure of a type `M` with respect to a set of type
+variables `T0, ..., Tn` to be `M` with every contravariant occurrence of `Ti`
+replaced with `Null`, and every covariant occurrence of `Ti` replaced with
+`Object`.
+
+TODO(leafp): Specify this precisely
+
+Note that the least closure of a type is a subtype of the original type (and of
+any substitution of types for the closed over type variables), and the greatest
+closure of a type is a supertype of the original type (and of any substitution
+of types for the closed over type variables).
+
+Discussion: We could consider taking bound information on the `Ti` into account,
+but this raises issues with respect to places where a default bound cannot be
+computed due to f-bounded subtyping.  If we do not take bounds into
+consideration (and even if we do) we may create types `A<Object>` where `Object`
+does not satisfy the bound for `A`.  If we accept super-bounded types of this
+form (as we have proposed to do) then this is not problematic.  If we don't
+accept these, then we need to be sure that there is eventually an appropriate
+error check on the final inferred types.
+
+## Upper bound
+
+We write `UP(T0, T1)` for the upper bound of `T0` and `T1`and `DOWN(T0, T1)` for
+the lower bound of `T0` and `T1`.  This extends to type schema by taking `UP(T,
+?) == T` and `DOWN(T, ?) == T` and symmetrically.
+
+TODO(leafp): write out the rules for this.
+
+## Subtype Constraints
+
+Subtype constraints are constraints on type variables `T` indicating that the
+constrained type variable is required to be either a subtype (`T <: M`) or a
+supertype (`M <: T`) of a specific type (or type schema) `M`.  Constraints in
+which the same variable occurs on the left and the right of the constraint are
+ill-formed.
+
+### Closure of subtype and supertype constraints.
+
+The closure of a subtype constraint `T <: M` with respect to a set of type
+variables `L` is the subtype constraint `T <: N` where `N` is the least closure
+of `M` with respect to `L`, and similarly for supertype constraints.
+
+The closure of a supertype constraint `M <: T` with respect to a set of type
+variables `L` is the subtype constraint `N <: T` where `N` is the greatest
+closure of `M` with respect to `L`, and similarly for supertype constraints.
+
+Note that the closure of a type constraint implies the original constraint: that
+is, any solution to the original constraint is a solution to the new constraint.
```

Comment:
I'm confused: is the closure of a type constraint intended to be more strict or less strict than the original constraint?  I think you intend it to be more strict, i.e. closure of (T <: U) with respect to U should be (T <: Null).  This is consistent with saying "the closure of a type constraint implies the original constraint" (i.e. (T <: Null) implies (T <: U)).

But it's not consistent with saying "any solution to the original constraint is a solution to the new constraint".  For the example above, "T=U" is a solution to the original constraint, but not to the new constraint.

Did you perhaps mean to say: "...that is, any solution to the new constraint is a solution to the original constraint."?